contracts/BufferLib.sol

```diff
@@ -12,6 +12,12 @@ library BufferLib {
     uint64 cursor;
   }
 
+  // Ensures we access an existing index in an array
+  modifier notOutOfBounds(uint256 index, uint256 length) {
+    require(index < length, "Array index out of bounds");
```

Comment:
I'd be surprised if a function called `readUint8` would fail with "array index out of bounds", given that neither the name or arguments of the function mention an array nor an index.

Maybe something like this? 
```suggestion
    require(index < length, "Tried to read from a consumed Buffer (must rewind it first)");
```